### PR TITLE
Cmdline refactor

### PR DESCRIPTION
**Pull request by nmathewson**

### CODE REVIEW

File: src/app/config/config.c [hunk 1]
```diff
@@ -2447,60 +2438,98 @@ options_act(const or_options_t *old_options)
...
+/**
+ * Enumeration to describe the syntax for a command-line option.
+ **/
 typedef enum {
-  TAKES_NO_ARGUMENT = 0,
+  /** Describe an option that does not take an argument. */
+  ARGUMENT_NONE = 0,
+  /** Describes an option that takes a single argument. */
   ARGUMENT_NECESSARY = 1,
+  /** Describes an option that takes a single optinal argument. */
```
Comment (teor2345) [inline src/app/config/config.c]: Typo:
```suggestion
  /** Describes an option that takes a single optional argument. */
```

File: src/app/config/tor_cmdline_mode.h [hunk 2]
```diff
@@ -16,14 +16,16 @@
...
-typedef enum tor_cmdline_mode_t {
+typedef enum {
   CMD_RUN_TOR=0, /**< The default: run Tor as a daemon. */
   CMD_LIST_FINGERPRINT, /**< Running --list-fingerprint. */
   CMD_HASH_PASSWORD, /**< Running --hash-password. */
   CMD_VERIFY_CONFIG, /**< Running --verify-config. */
   CMD_DUMP_CONFIG, /**< Running --dump-config. */
   CMD_KEYGEN, /**< Running --keygen */
   CMD_KEY_EXPIRATION, /**< Running --key-expiration */
+  CMD_OTHER, /**< Special value: indicates a command that is handled
```
Comment (teor2345) [inline src/app/config/tor_cmdline_mode.h]: Can we call this "CMD_IMMEDIATE"?
"OTHER" is not a very descriptive name.

File: src/app/config/config.c [hunk 3]
```diff
@@ -2459,6 +2459,8 @@ static const struct {
   takes_argument_t takes_argument;
   /** If not CMD_RUN_TOR, what should Tor do when it starts? */
   tor_cmdline_mode_t command;
+  /** If nonzero, set the quiet level to this. 1 is "hush", 2 is "quiet" */
```
Comment (teor2345) [inline src/app/config/config.c]: Can we please have this as a define or an enum?
I can never remember which is quieter :-)

File: src/app/config/config.c (hunk 1 above)
Comment (nmathewson) [reply src/app/config/config.c]: Fixup in 1b12c495f53972

File: src/app/config/config.c [hunk 4]
```diff
@@ -2459,8 +2450,12 @@ typedef enum {
   ARGUMENT_OPTIONAL = 2
 } takes_argument_t;
 
+/** Table describing arguments that Tor accepts on the command line,
```
[the remark this answers is not in the document]
Comment (nmathewson) [reply src/app/config/config.c]: ok, will leave alone for now. splitting commits is hard when I'm also adding fixups.

File: src/app/config/config.c [hunk 5]
```diff
@@ -2497,22 +2492,16 @@ static const struct {
...
 /** Helper: Read a list of configuration options from the command line.  If
- * successful, or if ignore_errors is set, put them in *<b>result</b>, put the
- * commandline-only options in *<b>cmdline_result</b>, and return 0;
- * otherwise, return -1 and leave *<b>result</b> and <b>cmdline_result</b>
- * alone. */
-int
-config_parse_commandline(int argc, char **argv, int ignore_errors,
-                         config_line_t **result,
-                         config_line_t **cmdline_result)
+ * successful, return a newly allocated parsed_cmdline_t; otherwise return
```
[the remark this answers is not in the document]
Comment (nmathewson) [reply src/app/config/config.c]: Fixup in b2f85badeac452

File: src/app/config/config.c [hunk 6]
```diff
@@ -5465,8 +5459,11 @@ options_init_from_string(const char *cf_defaults, const char *cf,
   }
 
   /* Go through command-line variables too */
-  retval = config_assign(get_options_mgr(), newoptions,
-                         global_cmdline_options, CAL_WARN_DEPRECATIONS, msg);
+  if (global_cmdline) {
```
[the remark this answers is not in the document]
Comment (nmathewson) [reply src/app/config/config.c]: I've done a fixup in e7f41383a33fb5 to try to make the intent more clear.

The reason I think the original was safe is that config_assign() has no effect when its config_line_t argument is NULL.

File: src/app/config/tor_cmdline_mode.h (hunk 2 above)
Comment (nmathewson) [reply src/app/config/tor_cmdline_mode.h]: Okay; done in 432791c58f5d01

**Pull request closed without merging**